### PR TITLE
[Flang][OpenMP] Fix comments that should not be Sentinels on fixed format.

### LINKED ISSUE #68653: [Flang][OpenMP] Compilation error by a comment line starting with "! $", "c $", or "* $" in the fixed source form
```
Version of flang-new : 18.0.0(cc627828f5176c6d75a25f1756d387d18539c1fb)
```

`! $`, `c $`, and `* $` (having a space before `$`) at the beginning of a line in fixed source form are treated as conditional compilation sentinels by mistake.

According to OpenMP 5.2 specification 3.3.1 (p.70)  
`Fixed Source Form Conditional Compilation Sentinels` have the following conditions:  
> The sentinel must start in column 1 and appear as a single word with no intervening white space. 

Therefore, these lines should be treated as comment lines.

The following are the test program, Flang-new, Gfortran and ifort compilation result.

ompfe1002_new2.f:
```fortran
      program main
! $   This is a comment line
C $   This is a comment line
c $   This is a comment line
* $   This is a comment line
      end program
```

```
$ flang-new ompfe1002_new2.f -fopenmp
error: Could not parse ompfe1002_new2.f
./ompfe1002_new2.f:2:5: warning: Character in fixed-form label field must be a digit
  ! $   This is a comment line
      ^
./ompfe1002_new2.f:3:5: warning: Character in fixed-form label field must be a digit
  C $   This is a comment line
      ^
./ompfe1002_new2.f:4:5: warning: Character in fixed-form label field must be a digit
  c $   This is a comment line
      ^
./ompfe1002_new2.f:5:5: warning: Character in fixed-form label field must be a digit
  * $   This is a comment line
      ^
./ompfe1002_new2.f:2:7: error: expected '('
  ! $   This is a comment line
        ^
./ompfe1002_new2.f:2:3: in the context: statement function definition
  ! $   This is a comment line
    ^
./ompfe1002_new2.f:2:3: in the context: declaration construct
  ! $   This is a comment line
    ^
./ompfe1002_new2.f:2:3: in the context: specification part
  ! $   This is a comment line
    ^
./ompfe1002_new2.f:1:7: in the context: main program
        program main
        ^
./ompfe1002_new2.f:3:7: error: expected '('
  C $   This is a comment line
        ^
./ompfe1002_new2.f:3:3: in the context: statement function definition
  C $   This is a comment line
    ^
./ompfe1002_new2.f:3:3: in the context: declaration construct
  C $   This is a comment line
    ^
./ompfe1002_new2.f:2:3: in the context: specification part
  ! $   This is a comment line
    ^
./ompfe1002_new2.f:1:7: in the context: main program
        program main
        ^
./ompfe1002_new2.f:4:7: error: expected '('
  c $   This is a comment line
        ^
./ompfe1002_new2.f:4:3: in the context: statement function definition
  c $   This is a comment line
    ^
./ompfe1002_new2.f:4:3: in the context: declaration construct
  c $   This is a comment line
    ^
./ompfe1002_new2.f:2:3: in the context: specification part
  ! $   This is a comment line
    ^
./ompfe1002_new2.f:1:7: in the context: main program
        program main
        ^
./ompfe1002_new2.f:5:7: error: expected '('
  * $   This is a comment line
        ^
./ompfe1002_new2.f:5:3: in the context: statement function definition
  * $   This is a comment line
    ^
./ompfe1002_new2.f:5:3: in the context: declaration construct
  * $   This is a comment line
    ^
./ompfe1002_new2.f:2:3: in the context: specification part
  ! $   This is a comment line
    ^
./ompfe1002_new2.f:1:7: in the context: main program
        program main
        ^
$
```

```
$ gfortran ompfe1002_new2.f -fopenmp
$
```

```
$ ifort ompfe1002_new2.f -qopenmp
$
```

### PR DESCRIPTION
**Pull request by Z572**
Fixes #68653

### CODE REVIEW

**Review by rofirrim — CHANGES_REQUESTED** [review]

File: flang/lib/Parser/prescan.cpp [hunk 1]
```diff
@@ -1174,18 +1174,19 @@ Prescanner::IsFixedFormCompilerDirectiveLine(const char *start) const {
   if (!IsFixedFormCommentChar(col1)) {
     return std::nullopt;
   }
+  if (p[0] != '$') {
```
[the remark this answers is not in the document]
Comment (Z572) [reply flang/lib/Parser/prescan.cpp]: looks like this code cannot diagnostic,i opened a new issues #68966

**Comment by kiranchandramohan** [comment]
> no idea why Lower/OpenMP/parallel-private-clause-fixes.f90 test fail.

Do you have https://github.com/llvm/llvm-project/commit/ef9f6170b61647e07ecbeabea7400bf4ac8e1fce in your branch from where you made this PR? If not could you rebase?

File: flang/test/Parser/OpenMP/sentinels.f [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+! CHECK:        end subroutine
+! CHECK:       end interface
+! CHECK:      PRINT *, "Is '    '"
+! CHECK:  123 PRINT *, "Is '123 '"
+
+!@cuf subroutine atcuf;
+      program main
+       interface
+        subroutine sub(a)
+!dir$ ignore_tkr a
```
Comment (rofirrim) [inline flang/test/Parser/OpenMP/sentinels.f:16]: Can you add a second dummy argument (say `b`) and split an `ignore_tkr` directive like so (mind the extra blank at the beginning of the continuing line), so we have a test for a continuation as well?

```fortran
        subroutine sub(a, b)
!dir$ ignore_tkr
!dir$+ b
!dir$ ignore_tkr a
      real(4):: a, b
```

File: flang/lib/Parser/prescan.cpp [hunk 3]
```diff
@@ -1186,6 +1186,8 @@ Prescanner::IsFixedFormCompilerDirectiveLine(const char *start) const {
         break;
       }
       *sp++ = ToLowerCaseLetter(*p);
+    } else {
```
Comment (rofirrim) [inline flang/lib/Parser/prescan.cpp:1189]: This is just a comment and no action is required.

A hypothetical line starting with `!$fp +` would not be considered a continuation of a previous line starting with `!$fp ` line as we check for `column == 6` below.

However, given that we do not have sentinels of length 2 or 3 (only 1 or 4, and length 1 is already handled in a special way) I believe this doesn't need solving now (I don't think one could write a test for that case currently).

**Review by rofirrim — COMMENTED** [review]
LGTM. But let's wait for @klausler in case he has further comments.

File: flang/lib/Parser/prescan.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rofirrim) [reply flang/lib/Parser/prescan.cpp]: I think with your last update, this is addressed.

File: flang/lib/Parser/prescan.cpp [hunk 4]
```diff
@@ -1186,6 +1186,8 @@ Prescanner::IsFixedFormCompilerDirectiveLine(const char *start) const {
...
+      break;
```
[the remark this answers is not in the document]
Comment (Z572) [reply flang/lib/Parser/prescan.cpp:1190]: `dir!` and `@cuf` can't contain Spaces, so I don't think they break the any existing code. and see https://github.com/llvm/llvm-project/pull/68911#discussion_r1357342950

**Comment by Z572** [comment]
ping

File: flang/lib/Parser/prescan.cpp (hunk 3 above)
Comment (klausler) [inline flang/lib/Parser/prescan.cpp:1189]: If the loop is now going to terminate on a space, then the various `break` conditions in the loop can be combined to be less confusing.

**Comment by Z572** [comment]
ping

**Review by klausler — APPROVED** [review]

**Pull request merged**